lib/wallet/src/Cardano/Wallet.hs

```diff
@@ -1635,15 +1636,64 @@ balanceTransaction ctx change pp ti wallet unadjustedPtx = do
                 @era @m @s @k @ktype
                 ctx change pp ti wallet strategy adjustedPtx
     balanceWith SelectionStrategyOptimal
-        `catchE` \case
-            ErrBalanceTxMaxSizeLimitExceeded
-                -> balanceWith SelectionStrategyMinimal
-            otherErr
-                -> throwE otherErr
+        `catchE` \e ->
+            if minimalStrategyIsWorthTrying e
+            then balanceWith SelectionStrategyMinimal
+            else throwE e
   where
     shelleyEra = Cardano.shelleyBasedEra @era
     recentEra = WriteTx.recentEra @era
 
+    -- Determines whether or not the minimal selection strategy is worth trying.
+    -- This depends upon the way in which the optimal selection strategy failed.
+    minimalStrategyIsWorthTrying :: ErrBalanceTx -> Bool
+    minimalStrategyIsWorthTrying e = or
+        [ maxSizeLimitExceeded
+        , unableToConstructChange
+        , selectionCollateralError
```

Comment:
Why not have one single `case e of`?

Reply:
But this is not set in stone: we can always revise this later!

Reply:
Sure! Ultimately/in general though, I suspect one `case` would be preferable because:

- A single `case` enables the compiler to warn against duplicate matches.
- We should simplify the error type anyway — it's so complicated! (I suspect a flat sum type without much nested sub-errors would be nice)

Reply:
> We should simplify the error type anyway — it's so complicated! (I suspect a flat sum type without much nested sub-errors would be nice)

Definitely! 💯 

Once we simplify the error type, I also think that moving to a single pattern match would be the obvious choice!